File: riku/core/svg_annotator.py
"""
Anotador de SVGs de Xschem.

Flujo:
1. Parsear el SVG para extraer posiciones de nombres de componentes.
2. Cruzar con el Schematic para calcular la transformacion .sch -> SVG.
3. Dibujar bounding boxes de color sobre componentes del DiffReport.
"""
import re
import xml.etree.ElementTree as ET
from dataclasses import dataclass

from riku.core.models import Schematic, DiffReport

_TEXT_RE = re.compile(
    r'<text[^>]+transform="translate\(([0-9.\-]+),\s*([0-9.\-]+)\)"[^>]*>\s*([^<]+?)\s*</text>'
)

# Color de nombre de componente en SVGs de Xschem — #cccccc
_COMPONENT_NAME_COLOR = "#cccccc"


@dataclass
class _Transform:
    mooz: float
    offset_x: float
    offset_y: float

    def to_svg(self, sch_x: float, sch_y: float) -> tuple[float, float]:
        return (
            sch_x * self.mooz + self.offset_x,
            sch_y * self.mooz + self.offset_y,
        )


def _extract_name_positions(svg_content: str) -> dict[str, tuple[float, float]]:
    """Extrae {nombre: (svg_x, svg_y)} de los textos de nombres de componentes."""
    positions: dict[str, tuple[float, float]] = {}
    for m in _TEXT_RE.finditer(svg_content):
        x, y, text = float(m.group(1)), float(m.group(2)), m.group(3).strip()
        # Verificar que el elemento es un nombre de componente por su color
        tag = m.group(0)
        if _COMPONENT_NAME_COLOR in tag:
            positions[text] = (x, y)
    return positions


def _fit_transform(
    svg_positions: dict[str, tuple[float, float]],
    schematic: Schematic,
) -> _Transform | None:
    """
    Calcula mooz, offset_x, offset_y usando minimos cuadrados sobre todos
    los componentes que aparecen tanto en el SVG como en el Schematic.
    Requiere al menos 2 puntos para resolver el sistema.
    """
    pairs = []
    for name, (svg_x, svg_y) in svg_positions.items():
        if name in schematic.components:
            comp = schematic.components[name]
            pairs.append((comp.x, comp.y, svg_x, svg_y))

    if len(pairs) < 2:
        return None

    # Minimos cuadrados para escala uniforme: svg = mooz * sch + offset
    # Separado por eje X e Y, luego promediamos mooz
    n = len(pairs)
    sum_sx = sum(p[0] for p in pairs)
    sum_sy = sum(p[1] for p in pairs)
    sum_vx = sum(p[2] for p in pairs)
    sum_vy = sum(p[3] for p in pairs)
    sum_sx2 = sum(p[0] ** 2 for p in pairs)
    sum_sy2 = sum(p[1] ** 2 for p in pairs)
    sum_sxvx = sum(p[0] * p[2] for p in pairs)
    sum_syvy = sum(p[1] * p[3] for p in pairs)

    denom_x = n * sum_sx2 - sum_sx ** 2
    denom_y = n * sum_sy2 - sum_sy ** 2

    if abs(denom_x) < 1e-9 or abs(denom_y) < 1e-9:
        return None

    mooz_x = (n * sum_sxvx - sum_sx * sum_vx) / denom_x
    mooz_y = (n * sum_syvy - sum_sy * sum_vy) / denom_y
    mooz = (mooz_x + mooz_y) / 2

    offset_x = (sum_vx - mooz * sum_sx) / n
    offset_y = (sum_vy - mooz * sum_sy) / n

    return _Transform(mooz=mooz, offset_x=offset_x, offset_y=offset_y)


_BBOX_HALF = 15  # mitad del lado del bounding box en unidades .sch

_COLORS = {
    "added":    ("rgba(0,200,0,0.25)",   "rgba(0,200,0,0.8)"),
    "removed":  ("rgba(200,0,0,0.25)",   "rgba(200,0,0,0.8)"),
    "modified": ("rgba(255,180,0,0.25)", "rgba(255,180,0,0.8)"),
}

_WIRE_STROKE = {
    "added":   "rgba(0,200,0,0.9)",
    "removed": "rgba(200,0,0,0.9)",
}
_WIRE_WIDTH = 2.5


def _wire_elements(wires, net_names: set[str], kind: str, transform: _Transform) -> list[str]:
    """Genera elementos <line> SVG para todos los wires que pertenecen a alguna net del set."""
    stroke = _WIRE_STROKE[kind]
    elements = []
    for w in wires:
        if w.label not in net_names:
            continue
        x1, y1 = transform.to_svg(w.x1, w.y1)
        x2, y2 = transform.to_svg(w.x2, w.y2)
        elements.append(
            f'<line x1="{x1:.2f}" y1="{y1:.2f}" x2="{x2:.2f}" y2="{y2:.2f}" '
            f'stroke="{stroke}" stroke-width="{_WIRE_WIDTH}" stroke-linecap="round"/>'
        )
    return elements


def annotate(
    svg_content: str,
    sch_b: Schematic,
    diff_report: DiffReport,
    sch_a: Schematic | None = None,
) -> str:
    """
    Recibe el SVG del commit B, los Schematics de ambos commits y el DiffReport.
    Retorna el SVG con:
    - Bounding boxes sobre componentes cambiados (added/removed/modified)
    - Trayectos de wires para nets añadidas (verde) y eliminadas (rojo)

    sch_a es opcional — si no se provee, las nets eliminadas no se dibujan.
    Si no se puede calcular la transformacion, retorna el SVG sin modificar.
    """
    svg_positions = _extract_name_positions(svg_content)
    transform = _fit_transform(svg_positions, sch_b)

    if transform is None:
        return svg_content

    elements = []
    half = _BBOX_HALF * transform.mooz

    # --- bounding boxes de componentes ---
    for cd in diff_report.components:
        # componentes added/modified usan sch_b; removed usan sch_a si disponible
        if cd.kind == "removed":
            source = sch_a if sch_a is not None else sch_b
        else:
            source = sch_b

        if cd.name not in source.components:
            continue
        comp = source.components[cd.name]
        cx, cy = transform.to_svg(comp.x, comp.y)
        fill, stroke = _COLORS.get(cd.kind, _COLORS["modified"])
        elements.append(
            f'<rect x="{cx - half:.2f}" y="{cy - half:.2f}" '
            f'width="{2*half:.2f}" height="{2*half:.2f}" '
            f'fill="{fill}" stroke="{stroke}" stroke-width="1.5" '
            f'rx="3" ry="3"/>'
        )

    # --- trayectos de wires de nets añadidas (en sch_b) ---
    if diff_report.nets_added:
        elements.extend(_wire_elements(
            sch_b.wires, set(diff_report.nets_added), "added", transform
        ))

    # --- trayectos de wires de nets eliminadas (en sch_a) ---
    if diff_report.nets_removed and sch_a is not None:
        elements.extend(_wire_elements(
            sch_a.wires, set(diff_report.nets_removed), "removed", transform
        ))

    if not elements:
        return svg_content

    annotation_layer = (
        '\n<g id="riku-diff-annotations">\n'
        + "\n".join(elements)
        + "\n</g>\n"
    )

    return svg_content.replace("</svg>", annotation_layer + "</svg>")

File: riku_rust/src/core/svg_annotator.rs
use std::collections::BTreeMap;
use std::path::Path;

use once_cell::sync::Lazy;
use regex::Regex;

use crate::core::models::{ChangeKind, DiffReport, Schematic};

static TEXT_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(
        r#"<text[^>]+transform="translate\(([0-9.\-]+),\s*([0-9.\-]+)\)"[^>]*>\s*([^<]+?)\s*</text>"#,
    )
    .expect("valid text regex")
});

static PATH_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"M([\d.\-]+) ([\d.\-]+)L([\d.\-]+) ([\d.\-]+)").expect("valid path regex"));

const COMPONENT_NAME_COLOR: &str = "#cccccc";
const BBOX_HALF: f64 = 15.0;

#[derive(Clone, Copy, Debug)]
struct Transform {
    mooz: f64,
    offset_x: f64,
    offset_y: f64,
    exact: bool,
}

impl Transform {
    fn to_svg(&self, sch_x: f64, sch_y: f64) -> (f64, f64) {
        (
            sch_x * self.mooz + self.offset_x,
            sch_y * self.mooz + self.offset_y,
        )
    }
}

fn change_style(kind: &ChangeKind, cosmetic: bool) -> (&'static str, &'static str) {
    match (kind, cosmetic) {
        (ChangeKind::Added, _) => ("rgba(0,200,0,0.25)", "rgba(0,200,0,0.8)"),
        (ChangeKind::Removed, _) => ("rgba(200,0,0,0.25)", "rgba(200,0,0,0.8)"),
        (ChangeKind::Modified, true) => ("rgba(120,120,120,0.20)", "rgba(120,120,120,0.85)"),
        (ChangeKind::Modified, false) => ("rgba(255,180,0,0.25)", "rgba(255,180,0,0.8)"),
    }
}

fn component_box(
    transform: Transform,
    component: &crate::core::models::Component,
    svg_position: Option<(f64, f64)>,
) -> (f64, f64, f64) {
    let (cx, cy) = svg_position.unwrap_or_else(|| transform.to_svg(component.x, component.y));
    let half = BBOX_HALF * transform.mooz;
    (cx, cy, half)
}

fn extract_name_positions(svg_content: &str) -> BTreeMap<String, (f64, f64)> {
    let mut positions = BTreeMap::new();
    for caps in TEXT_RE.captures_iter(svg_content) {
        let full_match = caps.get(0).map(|m| m.as_str()).unwrap_or("");
        if !full_match.contains(COMPONENT_NAME_COLOR) {
            continue;
        }
        let x = caps[1].parse::<f64>().unwrap_or(0.0);
        let y = caps[2].parse::<f64>().unwrap_or(0.0);
        let text = caps[3].trim().to_string();
        positions.insert(text, (x, y));
    }
    positions
}

fn extract_wire_endpoints(svg_content: &str) -> Vec<(f64, f64)> {
    let mut pts = Vec::new();
    for caps in PATH_RE.captures_iter(svg_content) {
        let x1 = caps[1].parse::<f64>().unwrap_or(0.0);
        let y1 = caps[2].parse::<f64>().unwrap_or(0.0);
        let x2 = caps[3].parse::<f64>().unwrap_or(0.0);
        let y2 = caps[4].parse::<f64>().unwrap_or(0.0);
        pts.push((x1, y1));
        pts.push((x2, y2));
    }
    pts
}

fn lstsq_free(pairs: &[(f64, f64, f64, f64)]) -> Option<Transform> {
    let n = pairs.len() as f64;
    let sum_sx: f64 = pairs.iter().map(|p| p.0).sum();
    let sum_sy: f64 = pairs.iter().map(|p| p.1).sum();
    let sum_vx: f64 = pairs.iter().map(|p| p.2).sum();
    let sum_vy: f64 = pairs.iter().map(|p| p.3).sum();
    let sum_sx2: f64 = pairs.iter().map(|p| p.0 * p.0).sum();
    let sum_sy2: f64 = pairs.iter().map(|p| p.1 * p.1).sum();
    let sum_sxvx: f64 = pairs.iter().map(|p| p.0 * p.2).sum();
    let sum_syvy: f64 = pairs.iter().map(|p| p.1 * p.3).sum();

    let denom_x = n * sum_sx2 - sum_sx * sum_sx;
    let denom_y = n * sum_sy2 - sum_sy * sum_sy;
    if denom_x.abs() < 1e-9 || denom_y.abs() < 1e-9 {
        return None;
    }

    let mooz_x = (n * sum_sxvx - sum_sx * sum_vx) / denom_x;
    let mooz_y = (n * sum_syvy - sum_sy * sum_vy) / denom_y;
    let mooz = (mooz_x + mooz_y) / 2.0;
    Some(Transform {
        mooz,
        offset_x: (sum_vx - mooz * sum_sx) / n,
        offset_y: (sum_vy - mooz * sum_sy) / n,
        exact: false,
    })
}

fn lstsq_fixed_origins(
    pairs: &[(f64, f64, f64, f64)],
    xorigin: f64,
    yorigin: f64,
    svg_wire_pts: Option<&Vec<(f64, f64)>>,
    sch_wire_pts: Option<&Vec<(f64, f64)>>,
) -> Option<Transform> {
    if let (Some(svg_wire_pts), Some(sch_wire_pts)) = (svg_wire_pts, sch_wire_pts) {
        let mut mooz_approx_x = Vec::new();
        for (sch_x, _sch_y, svg_x, _svg_y) in pairs {
            let denom = sch_x + xorigin;
            if denom.abs() > 1e-6 {
                mooz_approx_x.push(svg_x / denom);
            }
        }
        if !mooz_approx_x.is_empty() {
            let mooz_approx = mooz_approx_x.iter().sum::<f64>() / mooz_approx_x.len() as f64;
                let mut wire_pairs = Vec::new();
                for (sx, sy) in sch_wire_pts.iter().copied() {
                    let px = (sx + xorigin) * mooz_approx;
                    let py = (sy + yorigin) * mooz_approx;
                    if let Some(best) = svg_wire_pts.iter().min_by(|a, b| {
                    let da = (a.0 - px).powi(2) + (a.1 - py).powi(2);
                    let db = (b.0 - px).powi(2) + (b.1 - py).powi(2);
                    da.partial_cmp(&db).unwrap_or(std::cmp::Ordering::Equal)
                }) {
                    let dist = ((best.0 - px).powi(2) + (best.1 - py).powi(2)).sqrt();
                    if dist < 8.0 {
                        wire_pairs.push((sx, sy, best.0, best.1));
                    }
                }
            }
            if wire_pairs.len() >= 4 {
                let mut mooz_wx: Vec<f64> = wire_pairs
                    .iter()
                    .filter_map(|(schx, _schy, svgx, _svgy)| {
                        let denom = schx + xorigin;
                        if denom.abs() > 1e-6 {
                            Some(svgx / denom)
                        } else {
                            None
                        }
                    })
                    .collect();
                let mut mooz_wy: Vec<f64> = wire_pairs
                    .iter()
                    .filter_map(|(_schx, schy, _svgx, svgy)| {
                        let denom = schy + yorigin;
                        if denom.abs() > 1e-6 {
                            Some(svgy / denom)
                        } else {
                            None
                        }
                    })
                    .collect();

                if !mooz_wx.is_empty() && !mooz_wy.is_empty() {
                    for samples in [&mut mooz_wx, &mut mooz_wy] {
                        if samples.len() >= 4 {
                            let mean = samples.iter().sum::<f64>() / samples.len() as f64;
                            let variance = samples.iter().map(|v| (v - mean).powi(2)).sum::<f64>()
                                / samples.len() as f64;
                            let std = variance.sqrt();
                            samples.retain(|m| (*m - mean).abs() <= 2.0 * std);
                        }
                    }
                    if !mooz_wx.is_empty() && !mooz_wy.is_empty() {
                        let mooz = (mooz_wx.iter().sum::<f64>() / mooz_wx.len() as f64
                            + mooz_wy.iter().sum::<f64>() / mooz_wy.len() as f64)
                            / 2.0;
                        return Some(Transform {
                            mooz,
                            offset_x: xorigin * mooz,
                            offset_y: yorigin * mooz,
                            exact: true,
                        });
                    }
                }
            }
        }
    }

    let mut mooz_samples = Vec::new();
    for (sch_x, _sch_y, svg_x, _svg_y) in pairs {
        let denom = sch_x + xorigin;
        if denom.abs() > 1e-6 {
            mooz_samples.push(svg_x / denom);
        }
    }
    if mooz_samples.is_empty() {
        return None;
    }
    if mooz_samples.len() >= 4 {
        let mean = mooz_samples.iter().sum::<f64>() / mooz_samples.len() as f64;
        let variance =
            mooz_samples.iter().map(|v| (v - mean).powi(2)).sum::<f64>() / mooz_samples.len() as f64;
        let std = variance.sqrt();
        mooz_samples.retain(|m| (*m - mean).abs() <= 2.0 * std);
    }
    if mooz_samples.is_empty() {
        return None;
    }
    let mooz = mooz_samples.iter().sum::<f64>() / mooz_samples.len() as f64;
    Some(Transform {
        mooz,
        offset_x: xorigin * mooz,
        offset_y: yorigin * mooz,
        exact: true,
    })
}

fn fit_transform(
    svg_positions: &BTreeMap<String, (f64, f64)>,
    schematic: &Schematic,
    svg_path: Option<&Path>,
    svg_content: Option<&str>,
) -> Option<Transform> {
    let pairs: Vec<_> = svg_positions
        .iter()
        .filter_map(|(name, (svg_x, svg_y))| {
            schematic.components.get(name).map(|component| {
                (component.x, component.y, *svg_x, *svg_y)
            })
        })
        .collect();

    if pairs.len() < 2 {
        return None;
    }

    if let Some(svg_path) = svg_path {
        let origins_file = svg_path.parent()?.join("origins.txt");
        if origins_file.exists() {
            if let Ok(content) = std::fs::read_to_string(origins_file) {
                let mut lines = content.lines();
                if let (Some(xline), Some(yline)) = (lines.next(), lines.next()) {
                    if let (Ok(xorigin), Ok(yorigin)) = (xline.trim().parse::<f64>(), yline.trim().parse::<f64>()) {
                        let svg_wire_pts = svg_content.map(extract_wire_endpoints);
                        let sch_wire_pts: Option<Vec<(f64, f64)>> = if schematic.wires.is_empty() {
                            None
                        } else {
                            Some(
                                schematic
                                    .wires
                                    .iter()
                                    .flat_map(|w| [(w.x1, w.y1), (w.x2, w.y2)])
                                    .collect(),
                            )
                        };
                        if let Some(t) = lstsq_fixed_origins(
                            &pairs,
                            xorigin,
                            yorigin,
                            svg_wire_pts.as_ref(),
                            sch_wire_pts.as_ref(),
                        ) {
                            return Some(t);
                        }
                    }
                }
            }
        }
    }

    let t = lstsq_free(&pairs)?;
    if pairs.len() < 4 {
        return Some(t);
    }

    let residuals: Vec<f64> = pairs
        .iter()
        .map(|p| {
            let (x, y) = t.to_svg(p.0, p.1);
            (x - p.2).powi(2) + (y - p.3).powi(2)
        })
        .collect();
    let mean = residuals.iter().sum::<f64>() / residuals.len() as f64;
    let variance = residuals.iter().map(|r| (r - mean).powi(2)).sum::<f64>() / residuals.len() as f64;
    let std = variance.sqrt();
    let inliers: Vec<_> = pairs
        .iter()
        .zip(residuals.iter())
        .filter_map(|(pair, residual)| {
            if *residual <= mean + 2.0 * std {
                Some(*pair)
            } else {
                None
            }
        })
        .collect();
    if inliers.len() >= 2 {
        lstsq_free(&inliers)
    } else {
        Some(t)
    }
}

fn wire_elements(
    wires: &[crate::core::models::Wire],
    net_names: &std::collections::BTreeSet<String>,
    kind: &str,
    transform: Transform,
) -> Vec<String> {
    let stroke = match kind {
        "added" => "rgba(0,200,0,0.9)",
        "removed" => "rgba(200,0,0,0.9)",
        _ => "rgba(0,0,0,0.9)",
    };

    wires
        .iter()
        .filter(|w| net_names.contains(&w.label))
        .map(|w| {
            let (x1, y1) = transform.to_svg(w.x1, w.y1);
            let (x2, y2) = transform.to_svg(w.x2, w.y2);
            format!(
                r#"<line x1="{x1:.2}" y1="{y1:.2}" x2="{x2:.2}" y2="{y2:.2}" stroke="{stroke}" stroke-width="2.5" stroke-linecap="round"/>"#
            )
        })
        .collect()
}

pub fn annotate(
    svg_content: &str,
    sch_b: &Schematic,
    diff_report: &DiffReport,
    sch_a: Option<&Schematic>,
    svg_path: Option<&Path>,
) -> String {
    let svg_positions = extract_name_positions(svg_content);
    let transform = match fit_transform(&svg_positions, sch_b, svg_path, Some(svg_content)) {
        Some(t) => t,
        None => return svg_content.to_string(),
    };

    let mut elements = Vec::new();

    for cd in &diff_report.components {
        let source = if cd.kind == crate::core::models::ChangeKind::Removed {
            sch_a.unwrap_or(sch_b)
        } else {
            sch_b
        };
        if let Some(component) = source.components.get(&cd.name) {
            let (fill, stroke) = change_style(&cd.kind, cd.cosmetic);
            let (cx, cy, half) = component_box(transform, component, svg_positions.get(&cd.name).copied());
            elements.push(format!(
                r#"<rect x="{:.2}" y="{:.2}" width="{:.2}" height="{:.2}" fill="{fill}" stroke="{stroke}" stroke-width="1.5" rx="3" ry="3"/>"#,
                cx - half,
                cy - half,
                2.0 * half,
                2.0 * half
            ));
        }
    }

    if !diff_report.nets_added.is_empty() {
        let set: std::collections::BTreeSet<_> = diff_report.nets_added.iter().cloned().collect();
        elements.extend(wire_elements(&sch_b.wires, &set, "added", transform));
    }
    if let Some(sch_a) = sch_a {
        if !diff_report.nets_removed.is_empty() {
            let set: std::collections::BTreeSet<_> = diff_report.nets_removed.iter().cloned().collect();
            elements.extend(wire_elements(&sch_a.wires, &set, "removed", transform));
        }
    }

    if elements.is_empty() {
        return svg_content.to_string();
    }

    let annotation_layer = format!(
        "\n<g id=\"riku-diff-annotations\">\n{}\n</g>\n",
        elements.join("\n")
    );
    svg_content.replacen("</svg>", &(annotation_layer + "</svg>"), 1)
}

#[cfg(test)]
mod tests {
    use super::{annotate, change_style};
    use crate::core::models::{ChangeKind, Component, DiffReport, Schematic, Wire};
    use std::collections::{BTreeMap, BTreeSet};

    #[test]
    fn styles_cosmetic_changes_differently() {
        assert_eq!(
            change_style(&ChangeKind::Modified, true),
            ("rgba(120,120,120,0.20)", "rgba(120,120,120,0.85)")
        );
        assert_eq!(
            change_style(&ChangeKind::Modified, false),
            ("rgba(255,180,0,0.25)", "rgba(255,180,0,0.8)")
        );
    }

    #[test]
    fn annotate_injects_layer() {
        let mut components = BTreeMap::new();
        components.insert(
            "R1".to_string(),
            Component {
                name: "R1".to_string(),
                symbol: "res.sym".to_string(),
                params: BTreeMap::new(),
                x: 10.0,
                y: 20.0,
                rotation: 0,
                mirror: 0,
            },
        );
        components.insert(
            "R2".to_string(),
            Component {
                name: "R2".to_string(),
                symbol: "cap.sym".to_string(),
                params: BTreeMap::new(),
                x: 30.0,
                y: 10.0,
                rotation: 0,
                mirror: 0,
            },
        );
        let schematic = Schematic {
            components,
            wires: vec![Wire {
                x1: 0.0,
                y1: 0.0,
                x2: 10.0,
                y2: 0.0,
                label: "NET1".to_string(),
            }],
            nets: BTreeSet::from(["NET1".to_string()]),
        };
        let diff = DiffReport {
            components: vec![crate::core::models::ComponentDiff {
                name: "R1".to_string(),
                kind: ChangeKind::Added,
                cosmetic: false,
                before: None,
                after: None,
            }],
            nets_added: vec!["NET1".to_string()],
            nets_removed: vec![],
            is_move_all: false,
        };
        let svg = r##"<svg>
<text transform="translate(10, 10)" fill="#cccccc">R1</text>
<text transform="translate(30, 20)" fill="#cccccc">R2</text>
<path d="M0 0L10 0"/>
</svg>"##;

        let out = annotate(svg, &schematic, &diff, None, None);
        assert!(out.contains("riku-diff-annotations"));
        assert!(out.contains("<rect"));
        assert!(out.contains("<line"));
    }
}
